src/AccountManager.mo

```diff
@@ -491,40 +497,95 @@ module {
       };
     };
 
-    /// Initiates a withdrawal by transferring tokens to another account.
-    /// Returns ICRC1 transaction index and amount of transferred tokens (fee excluded).
-    public func withdraw(to : ICRC1.Account, amount : Nat) : async* WithdrawResponse {
-
-      totalWithdrawn_ += amount;
-
-      let callResult = await* processWithdrawTransfer(to, amount);
+    func handleWithdrawal(
+      from_subaccount : ?Blob,
+      to : ICRC1.Account,
+      amount : Nat,
+      onSuccess : (txIdx : Nat) -> async* WithdrawResponse,
+      onError : (err : ProcessWithdrawTransferError) -> async* WithdrawResponse,
+    ) : async* WithdrawResponse {
+      let callResult = await* processWithdrawTransfer(from_subaccount, to, amount);
 
       switch (callResult) {
-        case (#Ok txIdx) {
-          log(ownPrincipal, #withdraw({ to = to; amount = amount }));
-          #ok(txIdx, amount - fee(#withdrawal));
-        };
+        case (#Ok txIdx) { await* onSuccess(txIdx) };
         case (#Err(#BadFee { expected_fee })) {
           updateFee(expected_fee);
-          let retryResult = await* processWithdrawTransfer(to, amount);
+          let retryResult = await* processWithdrawTransfer(from_subaccount, to, amount);
           switch (retryResult) {
-            case (#Ok txIdx) {
+            case (#Ok txIdx) { await* onSuccess(txIdx) };
+            case (#Err err) { await* onError(err) };
+          };
+        };
+        case (#Err err) { await* onError(err) };
+      };
+    };
+
+    /// Initiates a direct withdrawal by transferring tokens to another account.
+    /// The transfer is made directly from the main account to the specified account indirectly.
+    /// In order to make the transfer traceable, the deposit account is used as an intermediary (fee x2).
+    /// Returns ICRC1 transaction index and amount of transferred tokens (fee excluded).
+    public func withdrawTraceably(p : Principal, to : ICRC1.Account, amount : Nat) : async* WithdrawResponse {
+      let ?release = depositRegistry.obtainLock(p) else return #err(#NotAvailable);
+
+      let depositAccount = {
+        owner = ownPrincipal;
+        subaccount = ?Util.toSubaccount(p);
+      };
+
+      let deposit = depositRegistry.get(p);
+
+      totalWithdrawn_ += amount;
+
+      await* handleWithdrawal(
+        null,
+        depositAccount,
+        amount,
+        func(txIdx : Nat) : async* WithdrawResponse {
+          await* handleWithdrawal(
+            ?Util.toSubaccount(p),
+            to,
+            amount,
+            func(txIdx : Nat) : async* WithdrawResponse {
+              ignore release(null);
               log(ownPrincipal, #withdraw({ to = to; amount = amount }));
-              #ok(txIdx, amount - fee(#withdrawal));
-            };
-            case (#Err err) {
+              #ok(txIdx, amount - 2 * fee(#withdrawal));
```

Comment:
This return value is the amount that was received by the destination `to` account, or?
The goal is to get `amount - fee(#withdrawal)` to the destination. We should have `2 * ledger_fee_ <= fee(#withdrawal)`, so `fee(#withdrawal)` needs to be subtracted only once.